api.go

```diff
@@ -583,6 +587,11 @@ func NewRaft(conf *Config, fsm FSM, logs LogStore, stable StableStore, snaps Sna
 // of them can be restored. This is called at initialization time, and is
 // completely unsafe to call at any other time.
 func (r *Raft) restoreSnapshot() error {
+	// Measure time from here since latency on listing snapshots or having to try
+	// multiple snapshots all adds to the restore time that matters i.e. how long
+	// the server takes to restart.
+	start := time.Now()
```

Comment:
I see what you are saying here, but I wonder how that will work out in practice. It seems strange that the first time we report this gauge it is going to include a bunch of other time that won't be reported by the other two sites which report this metric.

I imagine there is plenty of other "startup" time that is not reflected in this metric as well. Time required for the process to start, read config, and get to this point.

I wonder if it would be better to report this metric the same as the others (only around the fsm.Restore), and document that this is a minimum time and that a sufficient buffer should be added.

Or, maybe use `raft.fsm.restore` in all 3 call sites to indicate the time just for restore, and only report the gauge from this one call site? If the looping over other snapshots is going to be a significant amount of time, then reporting the gauge from other places and not including that time is going to report a much lower number and could be very misleading.

Reply:
That's fair. I reasoned that since this is still within the single logical operation of `restoreSnapshot` method it made some sense. 

Honestly I'm not super sure why we loop here - if the most recent completed snapshot on disk isn't readable for some reason then it's almost certain that we either don't have an older one, or if we do that we don't still have enough logs to restore from it since we would have truncated our logs after taking the newer one.

But yeah since it's an edge case I'm not sure is even possible it's not a big deal to move it back inside the loop. That will allow using a function/emitting in one place anyway as you suggested.